### PR TITLE
V1: Add conversion for Task.Resources

### PR DESCRIPTION
**Pull request by lbernick**
# Changes

This commit adds support for Task.Resources when converting between v1beta1 and v1
versions of Tasks. This will allow us to release v1 Task in a backwards compatible way,
by ensuring that v1beta1 Tasks with Resources will have the Resources serialized into
annotations on the v1 Task on conversion.

Part n of #4546 

/kind misc

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- n/a Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- n/a Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- n/a Release notes contains the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
NONE
```

### CODE REVIEW

**Comment by lbernick** [comment]
/retest

filed #5254

File: pkg/apis/version/conversion.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import (
+	"encoding/json"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// SerializeToMetadata serializes the input field and adds it as an annotation to
+// the metadata under the input key.
+func SerializeToMetadata(meta *metav1.ObjectMeta, field interface{}, key string) error {
```
Comment (JeromeJu) [inline pkg/apis/version/conversion.go:28]: - Shall we have tests for the two exported functions?
- Do we need to test with `v1.Task`?

Comment (lbernick) [reply pkg/apis/version/conversion.go:28]: > * Shall we have tests for the two exported functions?

Good point, thanks! Added a test.

> * Do we need test cases with `v1.Task` in task_conversion_tests?

No; here we are ensuring that a field of v1beta1 task that is not present in v1 is still there when converted to v1 and back. There's no equivalent for v1, since it's our highest version.

File: pkg/apis/version/version_validation_test.go [hunk 2]
```diff
@@ -14,19 +14,20 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package version
+package version_test
```
Comment (jerop) [inline pkg/apis/version/version_validation_test.go:17]: why are the changes in this file included in this commit? seems to me that it's an unrelated cleanup

Comment (lbernick) [reply pkg/apis/version/version_validation_test.go:17]: Split into a separate PR: https://github.com/tektoncd/pipeline/pull/5319

Comment (jerop) [reply pkg/apis/version/version_validation_test.go:17]: happy to lgtm this when these changes in this file are removed from this commit

Comment (lbernick) [reply pkg/apis/version/version_validation_test.go]: my bad-- for some reason I forgot it was possible to have some tests in a "version" package and some in a "version_test" package so I just assumed this would wait for 5319 to merge. done.
